### ISSUE TITLE
Add some Display Suite features to Manage Display page?

### ISSUE THREAD

**Issue by olafgrabienski** [backdrop/backdrop-issues#2084]
I'm considering to build new websites for some of my clients with Backdrop instead of Drupal. As I use [Display Suite](https://www.drupal.org/project/ds) a lot, I'm wondering if the module could be ported or if some of it's functionalities could be integrated into Backdrop core.

In my opinion, the main concept of Display Suite, i.e. the ability to arrange node displays (and other stuff) without touching template files, would fit perfectly the statement "Backdrop values site builders over coders". Reading layout related issues like #933 #774 or #1100 I got the impression that Backdrop members have reservations regarding Display Suite, which however might not be related to the mentioned Display Suite main concept but to other aspects of the Drupal module.

As I'm not able to code, I'd like to start a discussion if it makes sense to port some of the Display suite functionalities to Backdrop. Are there other (former) Drupal site builders who miss Display Suite in Backdrop? And as there are layouts in Backdrop, which Display suite features do you miss anyway?

One example: In #1100, @docwilmot stated that a meanwhile existing Backdrop feature (Adding blocks for adding individual fields from entities) "would replicate the functionality of Display Suite in Drupal: the ability to have fields from a node (or an entity) placed anywhere in a layout". The "adding field blocks" feature is a nice addition, but in my opinion the manage layout page shouldn't be the standard place to manipulate the display of nodes, and single blocks are not the ideal context to show node fields.

In summary, I'd like to manage the field display of nodes (or other entities) on the manage display pages, but the options are to less. Currently you can reorder fields there, hide or place the field labels and choose a format. I'd like to do some more stuff, e.g. change the name of the label, change the HTML output of the field, and some more stuff which coders might change in template files. What do you think about it?

**Comment by klonos** [comment]
I have only tested DS a couple of times as I did with Panels, so I can't say I miss something specific that these modules and their ecosystem offered, but I would like to...

a) point you to #779 as a related issue. @wesruv has provided some very promising designs there, but I know that being our main (perhaps sole) UI designer he has a lot on his plate - especially with the efforts to get the new front-end theme in core for 1.5.0. So I think that that issue has been put on hold for now.

b) I'd like to say that I envision Backdrop replacing the Field UI ("Manage fields" page) by implementing something similar to the Layout edit UI. It would allow users to build forms (content creation) using form layout templates that have header, footer and sidebar(s). Form regions would have "Add field" buttons instead of "Add block" buttons so that fields could be added using a modal. Regions would also have a "Configure" button that would allow users configuring them to be rendered as fieldsets, horizontal/vertical tabs, accordions etc ([Field Group](https://www.drupal.org/project/field_group) in core). Fields would be placed in the form regions as blocks that could be dragged around and also have "Configure links" that launch modals for editing their settings. I also imagine reusing that same layout/form UI to replace the "Manage display" page, but with fields now available as blocks in the Layout UI I think that perhaps we'd have overlapping functionality, so not sure about that. You say that you do not like how that (fields as blocks) currently works though. Hmm

**Comment by olafgrabienski** [comment]
Thanks for your feedback, @klonos, and thanks for pointing me to the promising issue #779. I like the ideas in the issue, and also your vision of replacing the Field UI by the Layout edit UI. Finally, I'm happy that you mentioned Field Group, as I use it together with Display Suite all the time.

Regarding the "fields as blocks" question: I see the existing approach (reuse blocks to display fields, managing them via Layouts) as a workaround, but I find it quite limiting regarding the HTML output. The output is too block-specific, and it's not really part of the node HTML. (Don't know if it's easy to change that behaviour.) Furthermore, the Manage layout page seems too 'far away' from the Field UI in the existing "fields as blocks" feature. (But as I understand your thoughts and the ideas in #779, that wouldn't be the case in the future.)

You guessed, that we should't expect quick progress regarding the Field UI issue. Do you think, it could be though a 1.x feature, or rather 2.x?

**Comment by wesruv** [comment]
@olafgrabienski I don't see these features as 'breaking', just additive. So long as that's the case, the only thing preventing us from getting it into 1.x is time/people power. These kinds of features are definitely what we're trying to work towards.

You are right in pointing out that these features are probably a little far away. 

I do think they can address all of the same problems that Display Suite was trying to solve. If you take a look at a custom block's settings and set it to "Dynamic", it'll let you set the markup of many (all?) of the elements in the block template, and lets you set classes on the element.

With that setup on layouts, manage display, fields as blocks, and having fieldgroups, I think we'll have replicated most of the DS features.

I'm also hoping we'll do something 'mini-panels'-esque in the long run, at which point I think we'll have all of my favorite features of panels and DS, but hopefully with a much much much better UI/UX

Re: Display Suite
I like a lot of the things the module is trying to do, but I don't like the implementation. I think that's mostly due to D7 core baggage, and I think Backdrop is in a good spot to solve the problems in a much nicer way. 

There might be opportunity to port specific features of DS in a more 'Backdroppy' way in contrib before we get stuff in core, and depending on implementation and how some of these efforts evolve the contrib work could be brought into core. (I've marked this as contrib candidate in hopes someone might take a stab at that :stuck_out_tongue:)

Right now I'm focused on the theme and responsive menus out of the box, but these kinds of features are high up on my list of things I want to do :smile:

**Comment by quicksketch** [comment]
One feature from DS that could (and should) be implemented as soon as possible is the ability to create custom display modes. We have an issue for this already at #315.

I'll add that some other features you brought up: Overriding the label text and specify customized HTML... these could be added as features in 1.x before or after any Field UI overhaul. Right now the ability to customize field display is pretty limited (just formatter settings). You're right that we want to allow for more control over markup through the UI and Field module lags behind Layout and Views modules in this regard. If we can pick out specific features and implement them individually, we can progressively iterate to add many of the DS features.

There are some items that don't particularly fit with Backdrop paradigms though. I've seen DS used to add layout-like structure (left/right columns etc) to the display of nodes. We'd probably want to implement a Layout-module driven solution to that sort of problem.

**Comment by olafgrabienski** [comment]
Thanks for your feedback! @quicksketch, I agree that only parts of DS are of interest while we don't want others like the DS layouts. As I use DS a lot I'd like to list the features which I find useful. May I do it here, in this issue?

**Comment by quicksketch** [comment]
> As I use DS a lot I'd like to list the features which I find useful. May I do it here, in this issue?

Sure! Though I've labeled this issue as a "question" for discussion rather than implementation. We'll need to split out these items into individual issues after we hash out what fits and what doesn't.

**Comment by olafgrabienski** [comment]
Fine! Before I move on, I'd like to list the DS like or related features which were mentioned positively in this issue so far:

- ability to create custom display modes
- arrange node displays without touching template files
- the ability to have fields from a node (or an entity) placed anywhere in a layout
- replace the Field UI by implementing something similar to the Layout edit UI
- fields placed in *form* regions as blocks that could be dragged around and have "Configure links" that launch modals for editing their settings
- build *forms* (for content creation) using form layout templates that have header, footer and sidebar(s)
- having fieldgroups: configure regions to be rendered as fieldsets, tabs, accordions etc
- something 'mini-panels'-esque but with better UI
- change the HTML output of fields / more control over markup through the UI
- set the markup of elements in the template via dynamic block settings
- setting classes on elements
- override the label text

Following I list the DS features which I use in most of my Drupal projects and which I would love to use in Backdrop. If you want to see my favorite features in action,

- install and enable DS in D7 (there is also a good working D8 version, different due to changes in D8)
- enable the DS Extras module and some features in it's settings (Field templates, Page title options, Regions to block)
- choose a DS layout for the view mode of your choice (even if you don't need it for layout purposes)

**Field templates:** Enabling the "extra" field templates, you can choose a different default template, for example "Full reset" which strips all HTML. I prefer "Minimal" which provides some markup including reasonable HTML classes but is less verbose than the Drupal core field template. Sometimes I use the option to override the chosen default template for individual fields by the "Expert" template which is fully customizable (and allows you to set classes). Apart from that, you can override the label text using every template. With Minimal and Expert you are also able to hide the colon at the end of inline labels, which is a very nice detail.

**Custom fields:** There is a UI to create four types of custom fields (code, dynamic, block, preprocess). In my opinion, code and block field are very useful in many cases. I had also use cases for dynamic fields which provide a CTools like UI with incredible much information related to the node and other system information. Not the right thing for Backdrop core, I guess!

**Title and other node information:** When you choose a DS layout (I use typically one column and not the more sophisticated ones), the Manage display page provides not only the fields of a content type but also node related information like Title, Post date, Author / User picture, Comments etc). You can drag them around and set them like fields. See the following screenshot as an example:
![ds-view-mode-full-7-title-and-other-node-information](https://cloud.githubusercontent.com/assets/14188307/18257749/43ead766-73c9-11e6-8fdd-d6a169bbb44e.png)

**Region as block:** Lets you create a region in the DS layout which will be provided as block and can be placed in the block region of your choice, let's say in a sidebar. Simple example use case: Content type book review, place the cover image and some book information in the sidebar.

**Last but not least, Field group:** Not part of DS but works together nicely and a must have in my opinion! Using both of them, you can accomplish everything you want.

What do you think, which of the mentioned features should go in Backdrop?

**Comment by olafgrabienski** [comment]
Related: https://github.com/backdrop/backdrop-issues/issues/2608
